flink/src/test/java/org/apache/iceberg/flink/TestRowDataPartitionKey.java

```diff
@@ -0,0 +1,258 @@
+    RowDataWrapper rowWrapper = new RowDataWrapper(FlinkSchemaUtil.convert(NESTED_SCHEMA), NESTED_SCHEMA.asStruct());
+    Iterable<RowData> rows = RandomRowData.generate(NESTED_SCHEMA, 10, 1991);
+
+    PartitionSpec spec1 = PartitionSpec.builderFor(NESTED_SCHEMA)
+        .identity("structType.innerStringType")
+        .build();
+    PartitionSpec spec2 = PartitionSpec.builderFor(NESTED_SCHEMA)
+        .identity("structType.innerIntegerType")
+        .build();
+
+    for (RowData row : rows) {
+      RowData innerRow = row.getRow(0, 2);
+
+      PartitionKey partitionKey1 = new PartitionKey(spec1, NESTED_SCHEMA);
+      partitionKey1.partition(rowWrapper.wrap(row));
+      Assert.assertEquals(partitionKey1.size(), 1);
+
+      String expectedStr = innerRow.isNullAt(0) ? null : innerRow.getString(0).toString();
+      Assert.assertEquals(expectedStr, partitionKey1.get(0, String.class));
+
+      PartitionKey partitionKey2 = new PartitionKey(spec2, NESTED_SCHEMA);
+      partitionKey2.partition(rowWrapper.wrap(row));
+      Assert.assertEquals(partitionKey2.size(), 1);
+
+      Integer expectedInt = innerRow.isNullAt(1) ? null : innerRow.getInt(1);
+      Assert.assertEquals(expectedInt, partitionKey2.get(0, Integer.class));
+    }
+  }
+
+  @Test
+  public void testPartitionMultipleNestedField() {
+    RowDataWrapper rowWrapper = new RowDataWrapper(FlinkSchemaUtil.convert(NESTED_SCHEMA), NESTED_SCHEMA.asStruct());
+    Iterable<RowData> rows = RandomRowData.generate(NESTED_SCHEMA, 10, 1992);
+
+    PartitionSpec spec1 = PartitionSpec.builderFor(NESTED_SCHEMA)
+        .identity("structType.innerIntegerType")
+        .identity("structType.innerStringType")
+        .build();
+    PartitionSpec spec2 = PartitionSpec.builderFor(NESTED_SCHEMA)
+        .identity("structType.innerStringType")
+        .identity("structType.innerIntegerType")
+        .build();
+
+    PartitionKey pk1 = new PartitionKey(spec1, NESTED_SCHEMA);
+    PartitionKey pk2 = new PartitionKey(spec2, NESTED_SCHEMA);
+
+    for (RowData row : rows) {
+      RowData innerRow = row.getRow(0, 2);
+
+      pk1.partition(rowWrapper.wrap(row));
+      Assert.assertEquals(2, pk1.size());
+
+      Integer expectedInt = innerRow.isNullAt(1) ? null : innerRow.getInt(1);
+      Assert.assertEquals(expectedInt, pk1.get(0, Integer.class));
+
+      String expectedStr = innerRow.isNullAt(0) ? null : innerRow.getString(0).toString();
+      Assert.assertEquals(expectedStr, pk1.get(1, String.class));
+
+      pk2.partition(rowWrapper.wrap(row));
+      Assert.assertEquals(2, pk2.size());
+
+      expectedStr = innerRow.isNullAt(0) ? null : innerRow.getString(0).toString();
+      Assert.assertEquals(expectedStr, pk2.get(0, String.class));
+
+      expectedInt = innerRow.isNullAt(1) ? null : innerRow.getInt(1);
+      Assert.assertEquals(expectedInt, pk2.get(1, Integer.class));
+    }
+  }
+
+  private static Object transform(Object value, Type type) {
```

Comment:
This has the same time bug as the getter. I think it would be better to avoid deriving expected values and just hard-code them. That makes tests easier to read, and more reliable.

Reply:
An alternative to hard-coding is to validate against a different object model. For example, you could generate the data with Iceberg generics, convert them to RowData, and then validate that the partitions produced from both object models match. That would catch the time bug.